### ISSUE TITLE
That `slicedim()` doesn't preserve the number of dimensions is sometimes frustrating

### ISSUE THREAD

**Issue by wsshin** [JuliaLang/julia#19708]
Currently, `slicedim(A, d, i)` takes the `i`th slice of a multi-dimensional array `A` in dimension `d`.  However, the number of  dimensions of the result is different from that of `A`:
```
julia> A = reshape(collect(1:8), (2,2,2))  # ndims = 3
2×2×2 Array{Int64,3}:
[:, :, 1] =
 1  3
 2  4

[:, :, 2] =
 5  7
 6  8

julia> slicedim(A, 2, 1)  # ndims = 2 ≠ 3
2×2 Array{Int64,2}:
 1  5
 2  6
```
This behavior generalizes Julia's behavior that `A[i, :]` returns a column vector rather than a row vector for a matrix `A`.  Initially this behavior confused me because in MATLAB `A[i,:]` returns a row vector.  However, the behavior has its own benefits (e.g., convenient when plotting a 2D slice of a 3D array because we don't have to reshape the slice into a 2D array), so I'm used to it now.

However, sometimes I badly want means for taking a slice without changing the number of dimensions.  For example, in my project I often need to take a slice of a multidimensional array and append the slice to the original array.  If `slicedim()` preserved the dimension, I could do it simply by `cat(d, A, slicedim(A, d, i))`.  However, because `slicedim()` changes the dimension, this code generates a `DimensionMismatch` error:
```
julia> cat(2, A, slicedim(A, 2, 1))
ERROR: DimensionMismatch("mismatch in dimension 3 (expected 2 got 1)")
 in _cs(::Int64, ::Bool, ::Int64, ::Int64) at ./abstractarray.jl:1129
 in _cshp at ./abstractarray.jl:1117 [inlined]
 in _cshp at ./abstractarray.jl:1126 [inlined] (repeats 2 times)
 in cat_shape(::Tuple{Bool,Bool}, ::Tuple{Int64,Int64,Int64}, ::Tuple{Int64,Int64}) at ./abstractarray.jl:1106 (repeats 2 times)
 in cat_t(::Int64, ::Type{T}, ::Array{Int64,3}, ::Vararg{Any,N}) at ./abstractarray.jl:1138
 in cat(::Int64, ::Array{Int64,3}, ::Array{Int64,2}) at ./abstractarray.jl:1244
```
To make it work, I need to reshape the result of `slicedim()`:
```
julia> cat(2, A, reshape(slicedim(A, 2, 1), (2,1,2)))
2×3×2 Array{Int64,3}:
[:, :, 1] =
 1  3  1
 2  4  2

[:, :, 2] =
 5  7  5
 6  8  6
```
This is inconvenient when the concatenation direction (`d = 2` in this case) is not known in advance, because the location of `1` in the tuple `(2,1,2)` indicating the size of the reshaped slice depends on `d`.

Of course, I can create the size to reshape the slice into by `s = [size(A)...]; s[d] = 1; s = (s...)` or `ntuple(k->(k==d ? 1: size(A)[k]), ndims(A))`.  However, it will be much more convenient if `slicedim()` takes an optional argument for preserving the dimension, like `slicedim(A, d, i[, squeeze = true])`.  Then, instead of `cat(d, A, reshape(slicedim(A, d, i), ntuple(k->(k==d ? 1: size(A)[k]), ndims(A))))`, I can do `cat(d, A, slicedim(A, d, i, false))`.

(Alternatively, it will be great if we have an opposite of `squeeze()`.  Suppose we have something like `insert_singletondims(A, dims)` that insert singleton dimensions to `A` at the dimensions specified by `dims`.  Then `cat(d, A, insert_singletondims(slicedim(A, d, i), d))` will do the job.)

Long story short, I think MATLAB's behavior of `A[:, ..., :, i, :, ..., :]` is more convenient than Julia's in some cases, like Julia's behavior is more convenient than MATLAB's in the other cases.  So I propose implementing some means to recover MATLAB's behavior, for example by adding an optional argument to `slicedim()` for preserving `ndims`.

**Comment by yuyichao** [comment]
`slicedim(A, 2, 1:1)`

**Comment by wsshin** [comment]
Wow, that's a very cool trick!  

So there was a reason the argument `i` of `slicedim(A, d, i)` was not `::Integer`.  I also found `slicedim(A, 2, [1])` works.

I also figured that `A[i:i, :]` and `A[[i], :]` returns a row vector for a matrix `A`!

I think this trick is so useful that it should be mentioned in the documentation for `slicedim()`!

**Comment by nalimilan** [comment]
The docs for `getindex` say `See the manual section on array indexing for details`. I guess a pull request adding the same mention for `slicedim` would be accepted.

**Comment by Sacha0** [comment]
Given how frequently this question comes up, making the solution highly visible (e.g. directly in the `getindex` and `slicedim` documentation) might be worthwhile?
